### DISCUSSION TITLE
In-game chat moderation tools

### DISCUSSION THREAD

**Discussion by peppy** [ppy/osu#27597] [Ideas]
Moving from an email with [Lyawi](https://osu.ppy.sh/users/5851253):

> I thought about the moderation-side of lazer in the future. And I don't know when it's time but I thought before you fully plan, you might be open to ideas too...? In case you're not, I am sorry then you can just ignore this mail
> I asked in osu!internal #gmt before, it seemed more fair to ask for their opinions first (you know, in case they would hate it in the first place). Not a lot of replies but they liked it.
> sooo, the "idea" is pretty much like Twitch moderation because it's efficient from what I've seen:
> 1.  **chat stops scrolling when you hover**
>     to prevent misclicking someone else
> 2.  **buttons next to username (delete, silence, copy message)**
>     a bit tricky if someone has a long username.. maybe?
>       Delete
>       This would be new. Delete message **without** a silence, maybe it should the user a pop-up warning for awareness (you could compare this to AFK warnings in other games). It's good for messages that are not appropriate in the chat but _maybe_ don't reeeeally deserve a silence?
>       Silence
>       The usual, maybe we could be able to actually choose the duration after clicking the button instead of stacking silences manually
>       Copy message
>       copies time, username, message. it's helpful
> 3.    recently silenced user appears differently in chat
>       It makes you aware the user might be a troublemaker. A red chat line background?
> 4.    **add notification in top right corner when you stack silences (duration of silenced user)**
>        In case we keep the stacking command silences, show a notification about silence current duration so that we wouldn't need to check ⁠lounge for it. That's how we do it atm and it's not really an issue, just a cool addition

**Comment by msmcculley** [comment]
This doesn't seem to have gotten much discussion despite lazer moderation tools (in-game) have not gotten any better. 

Iirc, the only thing that has been added to lazer that has any resemblance to moderation is the report feature that was graciously added by a community contributor (and GMT cannot use that report button for specific reasons). With my commenting on this to get the basic controls back to us. I don't speak for the team at all when I say this but I'd like to be able to play the game and the new features whilst still being able to moderate effectively from the game without needing to run an IRC at the same time. 

What I'm mainly looking for in with bringing this back to life is basic necessities at this point is the kill command, ways to disband multiplayer lobbies, and some other quality of life additions to it.

**Reply by peppy** [reply]
I think the best way forward is to bring ASS into chat, would you agree?

**Reply by msmcculley** [reply]
tl;dr: yes ASS commands are better than nothing. 

Some things I can foresee being useful:
> .silence (with a button please. writing out .s <"userid"> <"duration"> <"reason"> would be quite tenuous and would probably shy GMT/NAT away from using lazer moderation integration).

> #lounge or something similar to a report feed so we can see reports and act on them without needing to have another client or stable open. This would also need to log the silences in #lounge, if it doesn't already. Notifications of some sort could achieve this but I'd rather have a channel in which I can either see reports or see a logging of other moderators' (and myself) doings.

> If a full integration of ASS is to be used then .rm <avatar, cover, profile, userpage (not exactly needed due to lazer not having userpages up yet)> <-n or regular silence duration (42hr)> - This could be implemented on the userpage itself so it doesnt clutter up the chat line. Could also be used through a separate moderation channel with ASS or something.

> a way in ASS to unkill. There are times in which multiple GMT silence the user or NEED to silence themself for one reason or another, but mainly because of the stacking silences that doesnt warrant a stacking. 

Ideally, removing the option to report for GMT/NAT would be nice. Sometimes GMT report a user to get the user's logs and with the report button linking to the API/Web we cannot use that or !report (in lazer) in fear of restricting the user.

**Reply by peppy** [reply]
#lounge should already exist FWIW. lazer uses the same chat system

**Reply by msmcculley** [reply]
Currently, unless I am missing something, it is not able to be found: 

https://github.com/user-attachments/assets/558ed6c3-246a-4235-81aa-0171c6df8930

**Comment by tatitatai** [comment]
I tried to photoshop buttons next to names, it looked pretty bad but I'm also not good with that lol. 
If commands work like in stable, that would already be really awesome (faq and kill commands). As for #lounge, I agree it's not there.

**Comment by msmcculley** [comment]
There is going to be some talk in an upcoming gmt meeting (https://discord.com/channels/90072389919997952/1273031136339169280) to talk more indepth about what we would like to see implemented in lazer, of course dev presence for that might be good to figure out logistics and general feasibility of said additions and implementations.

**Reply by bdach** [reply]
I'm not sure I'll be able to be there (I don't think there's a time yet?) but I'd be hoping for meeting notes either way. Additionally the agenda seems quite wide, so I'm not sure whether a full call-response for feasibility of what you want to be in place for lazer wouldn't eat up the entire time of the meeting (I can't imagine this thing lasting more than like two-three hours tops, and there are seven large topics on the agenda?)

Probably just decide on what you want first and we can negotiate from there.

**Comment by 0x84f** [comment]
The GMT meeting concluded last week, and these are the bullet points of what would need to be done to achieve parity with stable, which would be a good start:
- #lounge needs to be accessible in-game
-  !kill, !unkill, !faq etc. need to be enabled/allowed (for lack of a better term) in all channels
- those commands also need to not show up in chat (discarded?)
- messages from silenced users should be removed from the chat

ASS can also be made available in the chat as already mentioned, but it would need the FAQ stuff moved there

_also aware of the current priority (or lack thereof) to implement this, as well as issue with access mentioned by spaceman_

**Reply by bdach** [reply]
Just gonna ping @ppy/team-web at least for awareness's sake since I imagine much of this will require web changes.

**Reply by nanaya** [reply]
I'm not sure `#2` and `#3` are possible with current system. the closest thing I can think of is redirecting known `!` commands to a dedicated http endpoint and return its result

`#4` is probably doable by fetching silenced user list from chat.updates endpoint and doing removal accordingly (although there's still problem when the silence period ends)

**Reply by bdach** [reply]
> `#2` and `#3` are possible with current system. the closest thing I can think of is redirecting known `!` commands to a dedicated http endpoint and return its result

I imagine folks would like these to work on web chat too. Not sure what that means in light of what you say.

> `#4` is probably doable by fetching silenced user list from chat.updates endpoint and doing removal accordingly

Do silences raise notifications or something, or would the client be forced to poll the silence list?

**Reply by nanaya** [reply]
> I imagine folks would like these to work on web chat too. Not sure what that means in light of what you say.

it means web chat needs to implement it as well...

> Do silences raise notifications or something, or would the client be forced to poll the silence list?

nope, web currently polls the silence list every now and then. there's plan to improve this a while ago but nothing has been done.

**Reply by bdach** [reply]
So I looked today and apparently we implement silences removing messages already...?

https://github.com/ppy/osu/blob/ea94f903c1ef72ef79dff19d923fdbd4bfcf2357/osu.Game/Online/Chat/ChannelManager.cs#L392-L402

Was even tested on production with assistance and it looked like it was working correctly, so not sure what it is about it not working.

**Comment by msmcculley** [comment]
Just touching base here. The GMT meeting happened today and lazer moderation was brought up. Nothing changed in terms of what was asked to be added, just wanted to make sure that this doesn't get lost in the sauce. 

Some of the things that are still missing that would be nice to have are: 
- Inherent way to silence a user on lazer. 
- #lounge is not added natively to lazer.
- Possibly adding FAQ functionality (i.e. !faq topic) 

Something that was brought up that isn't a necessity was: 
- The ability to purge a users chatlog or the entire channel chatlog without silencing users. 
--- This is particularly beneficial when users are being stupid with their posting (i.e. there have been instances in which users will post their financial information like credit card numbers, social security numbers, in real life addresses, etc.) and silencing the user is typically not deemed as appropriate so the gmt tend to kill then unkill but it's a process of two commands or more when there is a chain of users doing this behavior (which does happen).

**Reply by peppy** [reply]
Regarding this, I recently fixed a bug on bancho which stopped `!` commands from working in lazer/web. This was deployed today. Could you please give them a try and see if you have any success?

**Reply by msmcculley** [reply]
Hey peppy,

It looks as if the following commands work: 
`!k` works but is VISIBLE to everyone when done from lazer or web chat. 
`!report` works but is VISIBLE to everyone when done from lazer or web chat. 

The following commands do not work: 
`!faq` when done from a user with advanced permissions will be as if you are a normal player if done from lazer / web chat BUT will only get a DM from BanchoBot instead of posting the faq in the channel where its done (not exactly the same as what would happen as seen from IRC or Stable) 
`!` commands don't work in multiplayer
